### PR TITLE
Check if GEOPM_PLUGIN_PATH is secure

### LINKED ISSUE #2229: The IOGroup plugin path may be used by the service, but there is no check that the plugin path is secure
**Describe the bug**
If the GEOPM_PLUGIN_PATH environment variable is set in the geopm.service config file (this is not set in the default geopm.service unit), then it will be used by the service.  In this use case we must check that the directory is not writable by unprivileged users, otherwise the service may load IOGroup plugins which are not secure.  

**GEOPM version**
d538866

**Expected behavior**
Add a check when the plugin path is used such that if the calling process is privileged then the directory permissions are check prior to use.  

**Actual behavior**
No check is present, but the admin still must add an insecure directory to the geopm.service unit configuration file to create this problem. 

**Error messages produced**

**Additional context**

### PR DESCRIPTION
**Pull request by bgeltz**
- Fixes #2229.

### CODE REVIEW

File: service/geopmdpy/service.py [hunk 1]
```diff
@@ -57,6 +59,20 @@ def __init__(self):
...
 
+    def _check_plugins(self):
+        # Similar logic to geopm_plugin.cpp -> plugin_load()
+        so_suffix = '.so.' + version.__abi_version__.replace(':', '.')
+        plugin_path = os.getenv('GEOPM_PLUGIN_PATH')
+        if plugin_path is not None:
+            user_paths = plugin_path.split(':')
+            for path in user_paths:
+                if system_files.is_secure_path(path):
+                    for file_path in glob.glob(os.path.join(path, '*')):
```
Comment (cmcantalupo) [inline service/geopmdpy/service.py]: at least currently we also have .json files in the plugin directory.  We don't want to error when these .json files don't end in .so.1.0.0.

Comment (cmcantalupo) [inline service/geopmdpy/service.py]:
```suggestion
                    for file_path in glob.glob(os.path.join(path, 'lib*.so*')):
```

File: service/geopmdpy/service.py [hunk 2]
```diff
@@ -14,7 +14,6 @@
 import shutil
 import psutil
 import uuid
-import gi
```
Comment (bgeltz) [inline service/geopmdpy/service.py]: Was this important for any reason?  Seems unnecessary.

File: tutorial/iogroup/README.md [hunk 3]
```diff
@@ -197,6 +197,15 @@ filename must begin with "libgeopmiogroup_" and end in
 An alternative is to install the plugin by copying the .so file into
 the GEOPM install directory, <GEOPM_INSTALL_DIR>/lib/geopm.
 
+> [!WARNING]
```
Comment (bgeltz) [inline tutorial/iogroup/README.md:199]: GitHub markdown now supports admonition:
https://stackoverflow.com/a/72327818

Rendering:
https://github.com/bgeltz/geopm/tree/issue-2229/tutorial/iogroup

File: service/geopmdpy/system_files.py [hunk 4]
```diff
@@ -270,6 +271,42 @@ def is_secure_file(path, fid):
...
+
+    """
+    plugin_path = os.getenv('GEOPM_PLUGIN_PATH')
+    if plugin_path is not None:
+        user_paths = plugin_path.split(':')
+        for path in user_paths:
+            secure_make_dirs(path)
+            scanned_paths = set()
+            # Custom IOGroup/Agent plugins: geopm_plugin.cpp -> plugin_load()
+            for file_path in glob.glob(os.path.join(path, 'lib*.so*')):
```
Comment (cmcantalupo) [inline service/geopmdpy/system_files.py]: we have a more strict filter for loading plugins libgeopmpi*.so if i remember.

File: service/geopmdpy/system_files.py [hunk 5]
```diff
@@ -270,6 +271,42 @@ def is_secure_file(path, fid):
...
+                if not file_path.endswith('INVALID'):
+                    with open(file_path, 'r') as ff:
+                        scanned_paths.add(file_path)
+                        if is_secure_file(file_path, ff):
+                            so_suffix = '.so.' + version.__abi_version__.replace(':', '.')
```
Comment (cmcantalupo) [inline service/geopmdpy/system_files.py]: I guess these file checks are better than not having them, but there is a TOCTOU issue with this solution.

File: service/geopmdpy/system_files.py [hunk 6]
```diff
@@ -270,6 +271,42 @@ def is_secure_file(path, fid):
...
+    files that do not meet the permissions requirements will be renamed
+    such that they are ignored by the runtime.  Files that pass the check
+    will be loaded by the runtime.
+
+    """
+    plugin_path = os.getenv('GEOPM_PLUGIN_PATH')
+    if plugin_path is not None:
+        user_paths = plugin_path.split(':')
+        for path in user_paths:
+            secure_make_dirs(path)
```
Comment (cmcantalupo) [inline service/geopmdpy/system_files.py]: potentially moves files that are tracked by an RPM or deb file to /usr/lib64/geopm-INVALID, not sure if this is desired, as they will not be removed with an uninstall.

File: service/src/Helper.cpp [hunk 7]
```diff
@@ -269,6 +271,46 @@ namespace geopm
...
+        int err = 0;
+        struct stat stat_struct;
+        err = fstat(fd, &stat_struct);
+        if (err) {
+            throw Exception("fd_to_mode(): ", errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+        return stat_struct.st_mode;
+    }
+
+    bool is_fd_secure(const int fd, const std::string &plugin) {
```
Comment (cmcantalupo) [inline service/src/Helper.cpp]: maybe a more specific name is required for Helper.hpp, e.g. `is_plugin_secure`.

File: service/src/MSRIOGroup.cpp [hunk 8]
```diff
@@ -1260,7 +1261,20 @@ namespace geopm
...
-                    data_files.insert(filename);
+                    bool do_insert = true;
+                    int fd = open(filename.c_str(), O_RDONLY);
+                    if (fd < 0) {
+                        std::cerr << "Warning: <geopm> Failed to open() MSR plugin: "
+                                  << strerror(errno) << std::endl;
+                        do_insert = false;
+                    }
+                    else {
+                        // TODO: Insert fd into tracking vector to close on destruction
```
Comment (bgeltz) [inline service/src/MSRIOGroup.cpp]: This TODO is actually more difficult than I want it to be.  This function is static and is used in geopmadmin to generate the msr-safe allowlist.  Not sure if it's even worth it as the fd's will be close when the process ends.

File: service/geopmdpy/system_files.py [hunk 9]
```diff
@@ -910,6 +914,8 @@ def _get_names(exec_name):
                                    stdout=subprocess.PIPE,
                                    stderr=subprocess.PIPE)
     stdout_bytes, stderr_bytes = signals_pid.communicate()
```
Comment (cmcantalupo) [inline service/geopmdpy/system_files.py:916]: check return code for subprocess.

File: service/src/Helper.cpp [hunk 10]
```diff
@@ -269,6 +271,46 @@ namespace geopm
...
+        bool is_secure = true;
+        if (getuid() == 0) {
+            int uid = geopm::fd_to_uid(fd);
+            if (uid != 0) {
+                is_secure = false;
+                std::cerr << "Warning: <geopm> Plugin not owned by root and will be ignored: "
+                          << plugin << std::endl;
+            }
+            int mode = geopm::fd_to_mode(fd);
+            if (mode != (S_IFREG | S_IRWXU)) { // 0o700
```
Comment (cmcantalupo) [inline service/src/Helper.cpp]: Default plugins should have permissions 755 like other files in /usr/lib64.  So the validation here must be that the write bits are not set. For example 755 is okay, 700 is okay, but 777 is not okay

File: service/src/MSRIOGroup.cpp [hunk 11]
```diff
@@ -1256,13 +1257,25 @@ namespace geopm
         }
         std::vector<std::unique_ptr<MSR> > msr_arr_custom;
         for (const auto &dir : plugin_paths) {
-            auto files = list_directory_files(dir);
-            for (const auto &file : files) {
-                std::string filename = dir + "/" + file;
-                if (string_begins_with(file, "msr_") && string_ends_with(file, ".json")) {
-                    data_files.insert(filename);
+            try {
```
Comment (cmcantalupo) [inline service/src/MSRIOGroup.cpp:1266]:
```
class SecurePath {
    SecurePath(const std::string &orig_path)
    : SecurePath(orig_path, 0o022, getuid() == 0)
    {}
    SecurePath(const std::string &orig_path, int umask, bool is_same_user);
    ~SecurePath()
    std::string secure_path();
    int fid();
}
```

Comment (cmcantalupo) [reply service/src/MSRIOGroup.cpp:1266]: SecurePath sp("foobar");
// replace all uses of "foobar" with sp.secure_path()

File: service/src/SecurePath.cpp [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    SecurePath::SecurePath(const std::string &orig_path)
+        : SecurePath(orig_path,
+                     (S_IWGRP | S_IWOTH), // 0o022
+                     (getuid() == 0))
+    {
+
+    }
+
+    SecurePath::SecurePath(const std::string &orig_path, mode_t umask, bool enforce)
+        : m_fd(open(orig_path.c_str(), O_RDONLY))
```
Comment (bgeltz) [inline service/src/SecurePath.cpp]: Could add O_NOFOLLOW here if we want to disallow symlinks.

open() follows symlinks by default, and fstat cannot tell the difference between a link and a regular file when doing things this way.

File: service/src/geopm_plugin.cpp [hunk 13]
```diff
@@ -77,32 +77,41 @@ namespace geopm
...
-            }
-            else {
-                dl_handle = dlopen(plugin.c_str(), RTLD_LAZY|RTLD_GLOBAL);
+            try {
+                SecurePath sp (plugin.c_str());
+                void *dl_handle = dlopen(sp.secure_path().c_str(), RTLD_NOLOAD);
                 if (dl_handle != nullptr) {
                     DLRegistry::dl_registry().add(dl_handle);
                 }
-#ifdef GEOPM_DEBUG
```
Comment (bgeltz) [inline service/src/geopm_plugin.cpp:99]: Note I'm removing the ifdef.  I think it's important to print into the journal when plugins fail to load unexpectedly.

File: service/geopmdpy/system_files.py (hunk 9 above)
Comment (bgeltz) [reply service/geopmdpy/system_files.py:916]: Open an issue @bgeltz

File: service/docs/source/GEOPM_CXX_MAN_PluginFactory.3.rst [hunk 14]
```diff
@@ -206,6 +206,15 @@ detailed example of plugin load order can be found in
...
 
+.. warning::
+
+    It is imperative that any dynamically loaded library that is a dependency
+    of a plugin is available in the environment of the systemd initiated geopmd
+    service.  If dependent libararies are not available on the root user's
+    LD_LIBRARY_PATH, the plugin will fail to load.
+    For this reason, it is recommended that any plugins are compiled with
+    system installed gcc/g++ compilers.
+
```
Comment (bgeltz) [inline service/docs/source/GEOPM_CXX_MAN_PluginFactory.3.rst:218]: Add blurb about not allowing symlinks.

File: tutorial/iogroup/README.md [hunk 15]
```diff
@@ -197,6 +197,15 @@ filename must begin with "libgeopmiogroup_" and end in
...
+> When installing plugins system-wide for use through the GEOPM service,
+> it is imperative that any dynamically loaded library that is a dependency
+> of a plugin is available in the environment of the systemd initiated geopmd
+> service.  If dependent libararies are not available on the root user's
+> LD_LIBRARY_PATH, the plugin will fail to load.
+> For this reason, it is recommended that any plugins are compiled with
+> the system installed gcc/g++ compilers.
+
```
Comment (bgeltz) [inline tutorial/iogroup/README.md:207]: Add blurb about not allowing symlinks.

File: service/test/SecurePathTest.cpp [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+TEST_F(SecurePathTest, umask)
+{
+    mode_t test_perms = (S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH); // 0o644
+    chmod(m_file_name.c_str(), test_perms);
+
+    // Assert write permissions for the group/world are not set
+    SecurePath* sp = new SecurePath(m_file_name, (S_IWGRP | S_IWOTH), true);
+    EXPECT_EQ("/proc/self/fd/3", sp->secure_path());
+    delete sp;
```
Comment (bgeltz) [inline service/test/SecurePathTest.cpp]: Try with brackets instead of new/delete.

Comment (cmcantalupo) [inline service/test/SecurePathTest.cpp]:
```suggestion
    {
        SecurePath sp(m_file_name, (S_IWGRP | S_IWOTH), true);
        EXPECT_EQ("/proc/self/fd/3", sp.secure_path());
    }
```

File: service/src/SecurePath.cpp [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+            throw Exception("SecurePath::" + std::string(__func__) +
+                            "(): Failed to open file: ",
+                            errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        if (enforce) {
+            struct stat stat_struct;
+            int err = fstat(m_fd, &stat_struct);
+            if (err) {
+                (void)close(m_fd);
```
Comment (cmcantalupo) [inline service/src/SecurePath.cpp]:
```suggestion
                (void)!close(m_fd);
```
We need to put the ! in there with all these ignored return codes now

**Review by cmcantalupo — COMMENTED** [review]
Noticed the missing `!` from the ignored return values on this pass.  Also just wondering why the scoping trick doesn't enable us to avoid the new/delete pattern?

**Review by cmcantalupo — APPROVED** [review]

**Pull request merged**